### PR TITLE
checkpoint_abort not working when server is shutdown using qterm im…

### PR DESCRIPTION
**Pull request by prakashcv13**
…mediate and delay.

#### Bug/feature Description
* checkpoint_abort was not working when the server is shutdown using qterm -t immediate (and delay)
* This issue was found by running tests that are being reviewed under [https://github.com/PBSPro/pbspro/pull/749](PR#749)

#### Affected Platform(s)
* All

#### Cause / Analysis / Design
* checkpoint_abort was working in other scenarios.
* At the time of shutdown, the server was marking the substate of the job to be rerun before the MoM replied back on the checkpoint request.
* Due to this, in job_obit(), the flow was jumping to the rerun logic, which lead the server to send a delete job request.
#### Solution Description
* To not set the substate to rerun, the new state and substate will be evaluated and set in job_obit().

#### Testing logs/output
* Same 4 PTL test cases being reviewed in PR#749 were run. All test cases passed.
[ptl_0724.txt](https://github.com/PBSPro/pbspro/files/2224169/ptl_0724.txt)

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

**Comment by arungrover** [comment]
@prakashcv13 with this change now during server shutdown we will end up calling 'check_block' function in post job obit callback. Can you please run a test with and without your fix where there is a blocking job running on the system and see what happens when you run your test manually (one that reproduced the problem).
I suspect there might be some difference in behavior for blocking jobs.

**Comment by arungrover** [comment]
Thanks for confirming that there will be a behavior change. 
Since it is going to be visible to users, it qualifies as an external behavior change. 
It will attract a design change and a forum discussion on whether or not it is the right thing to do.

**Comment by prakashcv13** [comment]
@arungrover - I understand now what you were saying. The behavior change (calling check_block()) is due to the [PR-732](https://github.com/PBSPro/pbspro/pull/732). I will work on not calling check_block() for jobs that are being checkpointed.
@vccardenas - So far writing PTL with blocking jobs will block the test until the job gets over. I will work on using run_cmd to qsub a job with -Wblock.

**Comment by prakashcv13** [comment]
@arungrover, @lisa-altair, and @vcc - I have made the below changes -
1) Moved the call to check_block() after the checkpoint logic.
2) Changed PTL test cases so that they pass even if run by a user instead of root.
3) Checkpointed jobs on completion were getting re-queued as the checkpoint logic in job_obit() did not differentiate between Obit sent after checkpointing the job and the one sent after actual job completion, so added a new flag.
4) Set the new flag in req_holdjob() and shutdown_chkpt().
5) Reset the new flag when Obit is received for a hold job request.

All the test cases pass when run as a user and root, logs are attached.
[ptl.txt](https://github.com/PBSPro/pbspro/files/2284964/ptl.txt)
[ptl_user_logs.txt](https://github.com/PBSPro/pbspro/files/2284965/ptl_user_logs.txt)

Request you to review the changes.

File: src/server/req_holdjob.c [hunk 1]
```diff
@@ -210,6 +210,7 @@ req_holdjob(struct batch_request *preq)
 		} else {
 			pjob->ji_qs.ji_svrflags |=
 				JOB_SVFLG_HASRUN | JOB_SVFLG_CHKPT;
+			pjob->ji_qs.ji_svrflags |= JOB_SVFLG_HASHOLD;
```
Comment (arungrover) [inline src/server/req_holdjob.c]: I think this flag can move inside the previous statement itself.

I do have a question, You are adding this flag because job_obit() is unable to identify whether the job it is dealing with was checkpointed or has actually ended. But, my point of confusion is that why does the job holds on to JOB_SVFLG_CHKPT? When it is requeued the first time after checkpointing shouldn't we just remove CHKPT flag from the job.
I'm probably missing something here.

File: src/include/job.h [hunk 2]
```diff
@@ -891,6 +891,7 @@ task_find	(job		*pjob,
 #define JOB_SVFLG_ChkptMig 0x100 /* job has migratable checkpoint */
 #define JOB_SVFLG_Suspend  0x200 /* job suspended (signal suspend) */
 #define JOB_SVFLG_StagedIn 0x400 /* job has files that have been staged in */
+#define JOB_SVFLG_HASHOLD  0x800 /* job has a hold request sent to MoM */
```
Comment (arungrover) [inline src/include/job.h]: I'm surprised that you actually found a void in the job flags :)

File: src/server/req_shutdown.c [hunk 3]
```diff
@@ -323,9 +323,9 @@ shutdown_chkpt(job *pjob)
 		if (pjob->ji_qs.ji_state == JOB_STATE_TRANSIT)
 			svr_setjobstate(pjob, JOB_STATE_RUNNING, JOB_SUBSTATE_RUNNING);
 
-		pjob->ji_qs.ji_substate = JOB_SUBSTATE_RERUN;
```
Comment (lisa-altair) [inline src/server/req_shutdown.c]: Why did you remove setting the substate?

File: src/server/req_shutdown.c [hunk 4]
```diff
@@ -323,9 +323,9 @@ shutdown_chkpt(job *pjob)
...
 		pjob->ji_qs.ji_svrflags |= JOB_SVFLG_HASRUN;
 		pjob->ji_qs.ji_svrflags |= JOB_SVFLG_CHKPT;
+		pjob->ji_qs.ji_svrflags |= JOB_SVFLG_HASHOLD;
```
Comment (lisa-altair) [inline src/server/req_shutdown.c]: I think it would be best to OR all these flags together and then set ji_svrflags.

**Comment by prakashcv13** [comment]
@arungrover, @lisa-altair - thank you for the review. I have OR-ed the flags in one statement as per the review comments.

File: src/include/job.h (hunk 2 above)
Comment (lisa-altair) [reply src/include/job.h]: Yup!  Good idea.

File: src/server/req_shutdown.c (hunk 3 above)
Comment (lisa-altair) [reply src/server/req_shutdown.c]: Ah.  Okay, sorry I missed it.

**Comment by prakashcv13** [comment]
Thank you @lisa-altair for the sign-off.
@arungrover - could you please let me know if you have any further comments.

**Comment by prakashcv13** [comment]
Thank you @lisa-altair and @arungrover for the review and sign-off. I have squashed my commits and rebased my branch. 
@lisa-altair - as we decided, I have created [PP-1295](https://pbspro.atlassian.net/browse/PP-1295) for removing the duplicate flags.

**Comment by vccardenas** [comment]
@prakashcv13, the PTL tests look good to me.

**Comment by prakashcv13** [comment]
Thank you @vccardenas for the sign-off. I believe we can now take this PR forward.

**Comment by prakashcv13** [comment]
@PBSPro/pbspro-maintainers - could you please merge this PR.

**Comment by mike0042** [comment]
Please rebase your commit @prakashcv13.

**Comment by prakashcv13** [comment]
Thank you @mike0042, I just rebased.
